File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/util/TurnIcons.java
package net.cyclestreets.util;

import java.util.HashMap;
import java.util.Map;

import net.cyclestreets.view.R;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.support.v4.content.res.ResourcesCompat;

public final class TurnIcons
{
  public final static class Mapping  {
    private final Map<String, Drawable> mapping_;

    private Mapping(final Map<String, Drawable> mapping) {
      mapping_ = mapping;
    }

    public Drawable icon(final String turn) {
      final Drawable i = mapping_.get(turn.toLowerCase());
      return i != null ? i : mapping_.get("default");
    }
  }

  public static Mapping LoadMapping(final Context context) {
    return new Mapping(loadIconMappings(context));
  }

  private static Map<String, Drawable> loadIconMappings(final Context context) {
    final Resources res = context.getResources();

    return new HashMap<String, Drawable>() {{
        put("straight on", getDrawable(res, R.drawable.straight_on));
        put("bear left", getDrawable(res, R.drawable.bear_left));
        put("turn left", getDrawable(res, R.drawable.turn_left));
        put("sharp left", getDrawable(res, R.drawable.sharp_left));
        put("bear right", getDrawable(res, R.drawable.bear_right));
        put("turn right", getDrawable(res, R.drawable.turn_right));
        put("sharp right", getDrawable(res, R.drawable.sharp_right));
        put("double-back", getDrawable(res, R.drawable.double_back));
        put("join roundabout", getDrawable(res, R.drawable.roundabout));
        put("first exit", getDrawable(res, R.drawable.first_exit));
        put("second exit", getDrawable(res, R.drawable.second_exit));
        put("third exit", getDrawable(res, R.drawable.third_exit));
        put("waymark", getDrawable(res, R.drawable.waymark));
        put("default", getDrawable(res, R.drawable.ic_launcher));
      }};
  }

  private static Drawable getDrawable(Resources res, int iconId) {
    return ResourcesCompat.getDrawable(res, iconId, null);
  }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/util/TurnIcons.kt
package net.cyclestreets.util

import android.content.Context
import android.content.res.Resources
import android.graphics.drawable.Drawable
import androidx.core.content.res.ResourcesCompat
import net.cyclestreets.view.R

import net.cyclestreets.util.Turn.*

object TurnIcons {
    private lateinit var mapping: Map<Turn, Pair<Int, Drawable>>

    @JvmStatic
    fun initialise(context: Context) {
        val res = context.resources
        mapping = hashMapOf(
            STRAIGHT_ON to getDrawable(res, R.drawable.lr_straight_on),
            BEAR_LEFT to getDrawable(res, R.drawable.lr_bear_left),
            TURN_LEFT to getDrawable(res, R.drawable.lr_turn_left),
            SHARP_LEFT to getDrawable(res, R.drawable.lr_sharp_left),
            BEAR_RIGHT to getDrawable(res, R.drawable.lr_bear_right),
            TURN_RIGHT to getDrawable(res, R.drawable.lr_turn_right),
            SHARP_RIGHT to getDrawable(res, R.drawable.lr_sharp_right),
            TURN_LEFT_THEN_RIGHT to getDrawable(res, R.drawable.lr_turn_left_then_right),
            TURN_RIGHT_THEN_LEFT to getDrawable(res, R.drawable.lr_turn_right_then_left),
            BEAR_LEFT_THEN_RIGHT to getDrawable(res, R.drawable.lr_bear_left_then_right),
            BEAR_RIGHT_THEN_LEFT to getDrawable(res, R.drawable.lr_bear_right_then_left),
            DOUBLE_BACK to getDrawable(res, R.drawable.lr_double_back_right),
            FIRST_EXIT to getDrawable(res, R.drawable.lr_first_exit),
            SECOND_EXIT to getDrawable(res, R.drawable.lr_second_exit),
            THIRD_EXIT to getDrawable(res, R.drawable.lr_third_exit),
            WAYMARK to getDrawable(res, R.drawable.lr_waymark),
            DEFAULT to getDrawable(res, R.drawable.ic_launcher)
        )
    }

    @JvmStatic
    fun icon(turn: Turn): Drawable {
        return mappingFor(turn).second
    }

    @JvmStatic
    fun iconId(turn: Turn): Int {
        return mappingFor(turn).first
    }

    private fun mappingFor(turn: Turn): Pair<Int, Drawable> {
        return mapping[turn] ?: mapping[DEFAULT]!!
    }

    private fun getDrawable(res: Resources, iconId: Int): Pair<Int, Drawable> {
        return Pair(iconId, ResourcesCompat.getDrawable(res, iconId, null)!!)
    }
}
